### ISSUE TITLE
Constant output from exported ONNX

### ISSUE THREAD

**Issue by HudsonGraeme** [pytorch/pytorch#112512]
### 🐛 Describe the bug

After exporting a [NanoGPT](/karpathy/nanoGPT) torch model to ONNX, the outputs remain constant rather than changing based on `torch.rand` stochasticity and changes in input. Running the model in torch displays expected output; for the same input, the output is different each time, and the output appears to loosely match the provided input phrase. After export to ONNX, the model's output stays constant regardless of input.

I suspect the forward method contains logic that doesn't translate well over to ONNX, although haven't located a specific cause yet.

Here's a notebook setup with the entire flow of the model here (including training) but I can simplify further if required. In the notebook, the codeblock at the very bottom will display the issue when re-ran multiple times and with different outputs. The exported ONNX always remains constant no matter what changes are made to the `SAMPLE_TEXT` while the torch output changes each time.
> **NOTE**
> The notebook is designed to run on `mps`. You'll need to modify `device=` in the `Train the model` section if running on a non-mps device.

https://gist.github.com/HudsonGraeme/e397068bec442fce0628efa0a3709303

Things I've tried:
- Moving the `model.eval()` around within the export method
- Removing `model.eval()` entirely
- Adding `torch.compile` before exporting
- Switching `opset_versions` between 11 and 16 (needed to modify the model slightly for versions below 14 due to flash attention using dot product operations)
- Setting `do_constant_folding` to `False` in the ONNX export call
- Removing the `for` loop from `forward` entirely, and generating only one character from the model
- Removing the `torch.rand` call and replacing it with a constant float in the forward method
  > After this change the torch output stayed constant for a given input, but changed for different input as expected. The ONNX output still didn't change after modifying the input

Any help is appreciated!

### Versions

```
PyTorch version: 2.1.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 14.0 (arm64)
GCC version: Could not collect
Clang version: 15.0.0 (clang-1500.0.40.1)
CMake version: Could not collect
Libc version: N/A

Python version: 3.11.2 (main, Feb 28 2023, 11:34:36) [Clang 14.0.0 (clang-1400.0.29.202)] (64-bit runtime)
Python platform: macOS-14.0-arm64-arm-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Max

Versions of relevant libraries:
[pip3] numpy==1.26.1
[pip3] onnx==1.15.0
[pip3] onnxruntime==1.16.1
[pip3] torch==2.1.0
[conda] Could not collect
```

**Comment by HudsonGraeme** [comment]
While investigating it seems that `torch.rand()` is being treated as a constant when exported to ONNX. Not sure if this is expected behavior (I don't think it is). Here's a small repro of that:

https://gist.github.com/HudsonGraeme/372525c2fbf7a0fe42b8e22c99f7d0d5

**Comment by RuiVarela** [comment]
This does not seem to be a bug on pytorch or onnx,

nanGPT exports fine for onnx, the forward pass has no random steps in the flow, given the same input tokens it will produce the same output logits,

the random part of the flow is within the model sampling (python method generate), which is not exported, sampling is a post processing step, it's not part of the model.

here is both sampling implemented in python (the nanoGPT implementation) and also in javascript
https://github.com/RuiVarela/SmallLanguageModel 

https://github.com/RuiVarela/SmallLanguageModel/blob/main/web/inferencer.js

**Comment by thiagocrepaldi** [comment]
Thanks @RuiVarela 
Closing as it seems the exporter is working as expected
